snapcraft/internal/project_loader/grammar/_processor.py

```diff
@@ -134,6 +134,9 @@ def _parse_dict(
             if on_to_clause_match:
                 # We've come across the beginning of a compound statement
                 # with both 'on' and 'to'.
+                # The first time through this may be None, but the
+                # collection will ignore it.
+                statements.add(statement)
```

Comment:
why don't you just check for None?

Reply:
It's consistent with the remaining clauses (comment included).

I have some minor refactoring which takes care of this, but I wanted to push the bug fix first.

Reply:
I added a commit to do the refactor.